src/field/FieldDriver.hh

```diff
@@ -0,0 +1,88 @@
+//----------------------------------*-C++-*----------------------------------//
+// Copyright 2020 UT-Battelle, LLC, and other Celeritas developers.
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file FieldDriver.hh
+//---------------------------------------------------------------------------//
+#pragma once
+
+#include "base/Macros.hh"
+#include "base/Types.hh"
+
+#include "RungeKuttaStepper.hh"
+#include "MagFieldEquation.hh"
+#include "FieldParamsPointers.hh"
+#include "FieldInterface.hh"
+
+namespace celeritas
+{
+//---------------------------------------------------------------------------//
+/*!
+ * Brief class description.
+ *
+ *  This is a driver to control the quality of the field integration stepper
+ *  and provides the integration with a given field stepper
+ *
+ * \note This class is based on G4ChordFinder and G4MagIntegratorDriver
+ */
+class FieldDriver
+{
+  public:
+    // Construct with shared data and the stepper
+    inline CELER_FUNCTION
+    FieldDriver(const FieldParamsPointers&           shared,
+                RungeKuttaStepper<MagFieldEquation>& stepper);
+
+    // For a given trial step, advance by a sub_step within a tolerance error
+    CELER_FUNCTION real_type operator()(real_type step, OdeState* state);
+
+    // An adaptive step size control from G4MagIntegratorDriver
+    // Move this to private after all tests with non-uniform field are done
+    CELER_FUNCTION real_type accurate_advance(real_type step,
+                                              OdeState* state,
+                                              real_type hinitial);
+
+  private:
+    // A helper output for private member functions
+    struct FieldOutput
+    {
+        real_type step_taken; //!< Step length taken
+        real_type value;      //!< A helper variable
```

Comment:
Does it have a specific semantic that is still in need of finding a good name or different semantic depending of the functions?  If it is the later consider using an union to help distinguish and clarify if the caller and callee are really in sync, for example:
```suggestion
        union {  //!< A helper variable
            real_type proposed_step_length;
            real_type error_estimation;
            ... etc ... 
       };
```

Reply:
Good suggestion.  Implemented as suggested and updated codes accordingly.